### PR TITLE
Adding `Box` component in TS to component-library

### LINKED ISSUE #16976: 0 doesn't work for bounding box values for Box component
### Description
The value `0` doesn't work for box because there is a boolean check before running the `generateClassnames` function.

### LINKED ISSUE #19239: Create TS version of Box component in component-library
### Description
Currently the `Box` component is still a javascript component that lives in the `ui/components/ui` folder. This isn't very intuitive for engineers using the `Box` component as they have to import from 2 different places. It also effects CSS of other components because the specificity of the current Box component CSS is higher than the `ui/components/app` components

It would make sense to create a new TypeScript version of the `Box` component and deprecated the old one. That way engineers can import every component from the `component-library` folder and styles from the `Box` component won't override styles from the `app` components

### Technical Details
- Build a TypeScript version of the `Box` component in the `component-library` folder

### Acceptance Criteria
- There is a TypeScript version of the `Box` component in the `component-library` folder

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## Explanation
Currently, the `Box` component, a utility component used as a base for all `component-library` components as well as probably the most used component in the code base is still built in JavaScript. A types `box.d.ts` file was created as a workaround for allowing the `Box` component to be used in TypeScript files. However, through many community contributor TS migration PRs we have discovered that the polymorphic `as` prop is making all components difficult to migrate. This currently blocks all typescript migration and proper use of the Box component in TypeScript. 

To resolve this issue this PR creates a new `Box` component built in the `component-library` folder that:

-  ✅  Is built in TypeScript
-  ✅  Allows for correct typing of polymorphic props
-  ✅  Fixes https://github.com/MetaMask/metamask-extension/issues/16976
-  ✅  Allows all design system components to be imported from the same place e.g. `import { Box, Text, Button } from '../../component-library'`
- ✅  Unblocks [the TS migration of the rest of the `component-library` components](https://github.com/MetaMask/metamask-extension/issues/17725) as well as all extension TypeScript migration
- ✅  Improves the CSS specificity of the `Box` component to not override `ui/` and `app/` folder component styles like the current one does 

* Fixes #19239

### CSS Performance
This PR adds about `900kb` of extra CSS for the new `Box` component which I realize is A LOT. I've experimented using the same CSS for both components but that will either have some unexpected side effects if the old `Box` component uses the new CSS because of the updated specificity or we continue to have bad specificity for new `Box` component. I ran performance tests on css stylesheets and on average the increase in load time is around `79ms`. Performance tests should always be taken with a grain of salt because of background processes etc but overall I think it's not terrible and that we take this slight performance hit temporarily and actively start replacing the `Box` component through DS team and external contributor effort. Below is a table of 3 performance tests I ran on `develop` branch and this one.

| File             | Before (ms)              | Before Average (ms)     | After (ms)              | After Average (ms)     | Difference (ms)   |
|------------------|--------------------------|------------------------|-------------------------|------------------------|------------------:|
| index.css        | 214, 265, 301           | 260                    | 345, 301, 372          | 339.333                |       79.333      |
| index-rtl.css    | 267, 312, 581           | 386.667                | 306, 262, 329          | 299                    |      -87.667      |

Screenshot 2023-06-07 at 11 34 24 AM

Screenshot 2023-06-07 at 11 39 46 AM

## Next steps
Once this component has been merged we can deprecate the `Box` component in `ui/components/ui/box/box.js` and submit good first issues to the community to replace all instances of the old `Box` component inside the `component-library` folder and the rest of the code base. We can also update all the external contributor TS migration PRs

### Before
Storybook of `Box` component in `ui/components/ui/box/box.js`

https://github.com/MetaMask/metamask-extension/assets/8112138/b71e98c8-a930-4ee5-8dab-82a15082f5de

Shows polymorphic types not working 

Screenshot 2023-05-31 at 2 08 48 PM

### After
Storybook of `Box` component in `ui/components/component-library/box/box.tsx`

https://github.com/MetaMask/metamask-extension/assets/8112138/16d9028e-4a66-437d-82df-ebdc2284dc78

Shows polymorphic types working as expected and also erroring when not allowed

Screenshot 2023-05-31 at 2 07 04 PM

Testing Coverage

Screenshot 2023-06-05 at 8 52 59 PM

## Manual Testing Steps
Storybook testing
- Go to the latest build of storybook in this PR
- Go to the component-library version of the `Box` component
- Check all stories

Polymorphic `as` prop testing
- Pull this branch 
- Navigate to the `ui/components/component-library/box/box.stories.tsx` story
- Play around with the `as` prop changing the root element and testing different html attributes associated with the element passed to the `as` prop to check type checking

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/avatar-base/avatar-base.types.ts [hunk 1]
```diff
@@ -45,3 +46,10 @@ export interface AvatarBaseProps extends TextProps {
    */
   className?: string;
 }
+
+export type AvatarBaseProps<C extends React.ElementType> =
+  PolymorphicComponentPropWithRef<C, Props>;
+
+export type AvatarBaseComponent = <C extends React.ElementType = 'span'>(
+  props: AvatarBaseProps<C>,
+) => React.ReactElement | null;
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-base/avatar-base.types.ts:55]: Remove this

File: ui/components/component-library/avatar-base/avatar-base.tsx [hunk 2]
```diff
@@ -47,8 +52,8 @@ export const AvatarBase = forwardRef(
           className,
         )}
         ref={ref}
-        as={ValidTag.Div}
-        display={DISPLAY.FLEX}
+        as="div"
+        display={Display.Flex}
         justifyContent={JustifyContent.center}
         alignItems={AlignItems.center}
         borderRadius={BorderRadius.full}
```
Comment (georgewrmarshall) [inline ui/components/component-library/avatar-base/avatar-base.tsx:59]: Remove this

File: ui/components/component-library/button-icon/button-icon.scss [hunk 3]
```diff
@@ -11,7 +10,7 @@
   // ButtonIcon default states
   &:active,
   &:hover {
-    opacity: var(--button-icon-opacity-hover);
+    background-color: var(--color-background-alternative);
   }
 
   &--disabled,
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-icon/button-icon.scss:16]: Remove this

File: ui/components/component-library/index.js [hunk 4]
```diff
@@ -8,12 +8,13 @@ export { AvatarFavicon, AVATAR_FAVICON_SIZES } from './avatar-favicon';
...
 export {
   BadgeWrapper,
   BadgeWrapperPosition,
   BadgeWrapperAnchorElementShape,
 } from './badge-wrapper';
-export { AvatarBase } from './avatar-base';
+export { Box } from './box';
 export { Button, BUTTON_VARIANT, BUTTON_SIZES } from './button';
 export { ButtonBase, BUTTON_BASE_SIZES } from './button-base';
 export { ButtonIcon, ButtonIconSize } from './button-icon';
```
Comment (georgewrmarshall) [inline ui/components/component-library/index.js:20]: Remove this

File: ui/components/component-library/modal-header/modal-header.tsx [hunk 5]
```diff
@@ -58,7 +51,7 @@ export const ModalHeader: React.FC<ModalHeaderProps> = ({
     >
       {typeof children === 'string' ? (
         <Text
-          as={ValidTag.Header}
+          as="header"
           variant={TextVariant.headingSm}
           textAlign={TextAlign.Center}
         >
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-header/modal-header.tsx:57]: Remove this

File: ui/components/component-library/popover-header/popover-header.tsx [hunk 6]
```diff
@@ -57,7 +57,7 @@ export const PopoverHeader: React.FC<PopoverHeaderProps> = ({
         <Text
           variant={TextVariant.headingSm}
           textAlign={TextAlign.Center}
-          color={Color.inherit}
+          color={TextColor.inherit}
         >
           {children}
         </Text>
```
Comment (georgewrmarshall) [inline ui/components/component-library/popover-header/popover-header.tsx:63]: Remove this

File: ui/components/component-library/text/index.ts [hunk 7]
```diff
@@ -1,3 +1,3 @@
 export { Text } from './text';
 export { ValidTag, TextDirection, InvisibleCharacter } from './text.types';
-export type { TextProps } from './text.types';
+export type { TextStyleUtilityProps, TextProps } from './text.types';
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/index.ts:3]: Remove this

File: ui/components/component-library/text/text.test.tsx [hunk 8]
```diff
@@ -21,20 +21,20 @@ describe('Text', () => {
...
+        <Text as="span">span</Text>
+        <Text as="strong">strong</Text>
+        <Text as="em">em</Text>
+        <Text as="li">li</Text>
+        <Text as="div">div</Text>
+        <Text as="dt">dt</Text>
+        <Text as="dd">dd</Text>
       </>,
     );
     expect(container.querySelector('p')).toBeDefined();
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.test.tsx:40]: Remove this

File: ui/components/component-library/text/text.tsx [hunk 9]
```diff
@@ -1,84 +1,76 @@
...
+        as={tag}
+        dir={textDirection}
+        ref={ref}
+        {...props}
+      >
+        {children}
+      </Box>
+    );
+  },
+);
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.tsx:76]: This shows how the pattern for polymorphic props would work in the `Text` component there is a little more overhead to just extending `BoxProps`

File: ui/components/component-library/avatar-base/avatar-base.types.ts (hunk 1 above)
Comment (georgewrmarshall) [reply ui/components/component-library/avatar-base/avatar-base.types.ts]: updated

File: ui/components/component-library/box/box.types.ts [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+  C extends React.ElementType,
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  Props = {},
+> = PolymorphicComponentProp<C, Props> & { ref?: PolymorphicRef<C> };
+
+export interface StyleUtilityProps {
+  /**
+   * The content of the Box component.
+   */
+  children?: React.ReactNode;
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/box/box.types.ts]: Updated https://github.com/MetaMask/metamask-extension/pull/19363/commits/8ed1db2df54748826d83e2d00f69dbbb0a90ccf3

File: ui/components/component-library/box/box.types.ts [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+   * Use BackgroundColor enum from '../../../helpers/constants/design-system';
+   * Accepts responsive props in the form of an array.
+   */
+  backgroundColor?: BackgroundColor | BackgroundColorArray;
+  /**
+   * The text-color of the Box component.
+   * Use TextColor enum from '../../../helpers/constants/design-system';
+   * Accepts responsive props in the form of an array.
+   */
+  color?: TextColor | IconColor;
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/box/box.types.ts]: Updated https://github.com/MetaMask/metamask-extension/pull/19363/commits/8ed1db2df54748826d83e2d00f69dbbb0a90ccf3

File: ui/components/component-library/box/box.types.ts [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+];
+
+export type BackgroundColorArray = [
+  BackgroundColor,
+  BackgroundColor?,
+  BackgroundColor?,
+  BackgroundColor?,
+];
+
+export type PolymorphicRef<C extends React.ElementType> =
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/box/box.types.ts:184]: Updated https://github.com/MetaMask/metamask-extension/pull/19363/commits/8ed1db2df54748826d83e2d00f69dbbb0a90ccf3

File: ui/components/component-library/box/box.types.ts [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+   * The text-color of the Box component.
+   * Use TextColor enum from '../../../helpers/constants/design-system';
+   * Accepts responsive props in the form of an array.
+   */
+  color?: TextColor | IconColor;
+}
+
+interface Props extends StyleUtilityProps {
+  children: React.ReactNode;
+  className?: string;
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply ui/components/component-library/box/box.types.ts]: Updated https://github.com/MetaMask/metamask-extension/pull/19363/commits/8ed1db2df54748826d83e2d00f69dbbb0a90ccf3

**Review by darkwing — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]
This looks good to me!

**Review by garrettbear — APPROVED** [review]

**Pull request merged**